controllers/nnf_port_manager_controller_test.go

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"fmt"
+	"reflect"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	dwsv1alpha1 "github.com/HewlettPackard/dws/api/v1alpha1"
+	nnfv1alpha1 "github.com/NearNodeFlash/nnf-sos/api/v1alpha1"
+)
+
+var _ = Context("NNF Port Manager Controller Setup", Ordered, func() {
+
+	var cfg *dwsv1alpha1.SystemConfiguration
+	const portStart = 20
+	const portEnd = 29
+
+	BeforeAll(func() {
+		cfg = &dwsv1alpha1.SystemConfiguration{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "port-manager-system-config",
+				Namespace: corev1.NamespaceDefault,
+			},
+			Spec: dwsv1alpha1.SystemConfigurationSpec{
+				Ports: []intstr.IntOrString{
+					intstr.FromString(fmt.Sprintf("%d-%d", portStart, portEnd)),
+				},
+			},
+		}
+
+		Expect(k8sClient.Create(ctx, cfg)).To(Succeed())
+		DeferCleanup(func() { Expect(k8sClient.Delete(ctx, cfg)).To(Succeed()) })
+	})
+
+	Describe("NNF Port Manager Controller Test", func() {
+
+		var mgr *nnfv1alpha1.NnfPortManager
+		var r = &NnfPortManagerReconciler{} // use this to access private reconciler methods
+
+		BeforeEach(func() {
+			mgr = &nnfv1alpha1.NnfPortManager{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "nnf-port-manager",
+					Namespace: corev1.NamespaceDefault,
+				},
+				Spec: nnfv1alpha1.NnfPortManagerSpec{
+					SystemConfiguration: corev1.ObjectReference{
+						Name:      cfg.Name,
+						Namespace: cfg.Namespace,
+						Kind:      reflect.TypeOf(*cfg).Name(),
+					},
+					Allocations: make([]nnfv1alpha1.NnfPortManagerAllocationSpec, 0),
+				},
+			}
+		})
+
+		JustBeforeEach(func() {
+			Expect(k8sClient.Create(ctx, mgr)).To(Succeed())
+			DeferCleanup(func() { Expect(k8sClient.Delete(ctx, mgr)).To(Succeed()) })
+		})
+
+		reservePorts := func(mgr *nnfv1alpha1.NnfPortManager, name string, count int) {
+			By(fmt.Sprintf("Reserving %d ports for '%s'", count, name))
+
+			allocation := nnfv1alpha1.NnfPortManagerAllocationSpec{
+				Requester: corev1.ObjectReference{Name: name},
```

Comment:
Normally, the reference object would have to actually exist, right?

Reply:
I'm not sure I understand your question. As long as you can make an ObjectReference{}, it will work. Create one out of thin air if you want.

Reply:
Ah. Initially I was thinking it has to reference an object that exists.